cranelift/codegen/src/isa/x64/inst/args.rs

```diff
@@ -0,0 +1,434 @@
+    pub(crate) fn get_regs_as_uses(&self, collector: &mut RegUsageCollector) {
+        match self {
+            RMI::R { reg } => collector.add_use(*reg),
+            RMI::M { addr } => addr.get_regs_as_uses(collector),
+            RMI::I { simm32: _ } => {}
+        }
+    }
+}
+
+impl ShowWithRRU for RMI {
+    fn show_rru(&self, mb_rru: Option<&RealRegUniverse>) -> String {
+        self.show_rru_sized(mb_rru, 8)
+    }
+
+    fn show_rru_sized(&self, mb_rru: Option<&RealRegUniverse>, size: u8) -> String {
+        match self {
+            RMI::R { reg } => show_ireg_sized(*reg, mb_rru, size),
+            RMI::M { addr } => addr.show_rru(mb_rru),
+            RMI::I { simm32 } => format!("${}", *simm32 as i32),
+        }
+    }
+}
+
+/// An operand which is either an integer Register or a value in Memory.  This can denote an 8, 16,
+/// 32 or 64 bit value.
+#[derive(Clone)]
+pub(crate) enum RM {
+    R { reg: Reg },
+    M { addr: Addr },
+}
+
+impl RM {
+    // Constructors.
+
+    fn reg(reg: Reg) -> Self {
+        debug_assert!(reg.get_class() == RegClass::I64);
+        RM::R { reg }
+    }
+
+    fn mem(addr: Addr) -> Self {
+        RM::M { addr }
+    }
+
+    /// Add the regs mentioned by `self` to `collector`.
+    pub(crate) fn get_regs_as_uses(&self, collector: &mut RegUsageCollector) {
+        match self {
+            RM::R { reg } => collector.add_use(*reg),
+            RM::M { addr } => addr.get_regs_as_uses(collector),
+        }
+    }
+}
+
+impl ShowWithRRU for RM {
+    fn show_rru(&self, mb_rru: Option<&RealRegUniverse>) -> String {
+        self.show_rru_sized(mb_rru, 8)
+    }
+
+    fn show_rru_sized(&self, mb_rru: Option<&RealRegUniverse>, size: u8) -> String {
+        match self {
+            RM::R { reg } => show_ireg_sized(*reg, mb_rru, size),
+            RM::M { addr } => addr.show_rru(mb_rru),
+        }
+    }
+}
+
+#[derive(Clone, PartialEq)]
+// Some basic ALU operations.  TODO: maybe add Adc, Sbb.
+pub enum RMI_R_Op {
+    Add,
+    Sub,
+    And,
+    Or,
+    Xor,
+    Mul, // the signless, non-extending (N x N -> N, for N in {32,64}) variant
+}
+
+impl RMI_R_Op {
+    pub(crate) fn to_string(&self) -> String {
+        match self {
+            RMI_R_Op::Add => "add".to_string(),
+            RMI_R_Op::Sub => "sub".to_string(),
+            RMI_R_Op::And => "and".to_string(),
+            RMI_R_Op::Or => "or".to_string(),
+            RMI_R_Op::Xor => "xor".to_string(),
+            RMI_R_Op::Mul => "imul".to_string(),
+        }
+    }
+}
+
+impl fmt::Debug for RMI_R_Op {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        write!(fmt, "{}", self.to_string())
+    }
+}
+
+#[derive(Clone, PartialEq)]
+// These indicate ways of extending (widening) a value, using the Intel
+// naming: B(yte) = u8, W(ord) = u16, L(ong)word = u32, Q(uad)word = u64
+pub enum ExtMode {
+    BL, // B -> L
+    BQ, // B -> Q
+    WL, // W -> L
+    WQ, // W -> Q
+    LQ, // L -> Q
```

Comment:
```suggestion
    /// byte -> long word
    BL,
    /// byte -> quad word
    BQ,
    /// word -> long word
    WL,
    /// word -> quad word
    WQ,
    /// long -> quad word
    LQ,
```

Also is there no byte -> word variant?

Reply:
encoding-wise, `movzx` and `movsx` have such an encoding with `660f{b6,be}`. For example, `movzx ax, cl // 66 0f b6 c1`. I dunno if we're just not using them though.

Reply:
I propose not to generate 16 bit operations (addw, etc) to the extent that it's possible to avoid them.  They put us at risk of partial register stalls.  Instead I plan to do as much 16-bit work as possible at 32 bit width, hence avoiding the stalls and also the need for the 66 prefix.

Reply:
that sounds reasonable to me (pending if/when we might also optimize for size at least)

Do you know if decode stalls from 66-prefixed instructions with immediates are still a concern? At the least it might be helpful to note what is _intentionally_ not implemented, for future readers.

Reply:
I don't know about decode stalls offhand; we'd have to consult the latest incarnation of the Intel "Optimisation Reference Manual".  Note though that I was speaking above of dispatch stalls, not decode stalls -- I forgot there might also be decode limitations till you mentioned it.

Yes, documenting the general intention to avoid generating 16-bit operations as much as possible would be a good thing to do.